lib/common/rules.c

```diff
@@ -605,64 +678,32 @@ pcmk__evaluate_date_expression(const xmlNode *date_expression,
  * \param[in] nmatches    Number of entries in \p submatches[]
  *
  * \return Newly allocated string identical to \p string with submatches
- *         expanded, or NULL if there were no matches
+ *         expanded on success, or NULL if no expansions were needed
+ * \note The caller is responsible for freeing the result with free()
  */
 char *
 pcmk__replace_submatches(const char *string, const char *match,
                          const regmatch_t submatches[], int nmatches)
 {
-    size_t len = 0;
-    int i;
-    const char *p, *last_match_index;
-    char *p_dst, *result = NULL;
+    size_t nbytes = 0;
+    char *result = NULL;
 
     if (pcmk__str_empty(string)) {
-        return NULL;
+        return NULL; // Nothing to expand
     }
 
-    p = last_match_index = string;
-
-    while (*p) {
-        if (*p == '%' && *(p + 1) && isdigit(*(p + 1))) {
-            i = *(p + 1) - '0';
-            if ((nmatches >= i) && (submatches[i].rm_so != -1)
-                && (submatches[i].rm_eo > submatches[i].rm_so)) {
-                len += p - last_match_index
-                       + (submatches[i].rm_eo - submatches[i].rm_so);
-                last_match_index = p + 2;
-            }
-            p++;
-        }
-        p++;
+    // Calculate how much space will be needed for expanded string
```

Comment:
I think we can simplify this a lot by building up a `GString`, instead of doing one loop to compute the size and another loop to do the expansion.

Simply don't allocate a `GString` until we find an expansion. I wouldn't worry about reallocations with a starting size like 256... should be FAR more than enough for almost any real use case.

Requires only minor refactoring at the two callers to handle a `gchar *` or `GString *`.

Reply:
I think you're right, but I'm tired of looking at this code :) and I think it's reasonable to follow now. At least being exact with the memory is a little benefit.